File: app/schemas/history.py
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel

HistoryJobType = Literal["batch", "priority"]


class ClientHistoryItem(BaseModel):
    request_id: int
    delivery_type: HistoryJobType
    request_status: str

    volume_liters: int
    created_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    batch_id: Optional[int] = None
    member_id: Optional[int] = None
    batch_status: Optional[str] = None
    member_status: Optional[str] = None
    payment_status: Optional[str] = None
    refund_status: Optional[str] = None
    amount_paid: Optional[float] = None

    tanker_id: Optional[int] = None
    driver_name: Optional[str] = None

    delivery_id: Optional[int] = None
    delivery_status: Optional[str] = None
    planned_liters: Optional[float] = None
    actual_liters_delivered: Optional[float] = None
    otp_verified: Optional[bool] = None
    delivered_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ClientHistoryResponse(BaseModel):
    user_id: int
    total: int
    items: list[ClientHistoryItem]


class DriverHistoryItem(BaseModel):
    job_type: HistoryJobType
    job_id: int

    tanker_id: int
    tanker_status: Optional[str] = None

    total_stops: int
    delivered_stops: int
    failed_stops: int
    skipped_stops: int

    total_planned_liters: float
    total_actual_liters_delivered: float

    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    last_updated_at: Optional[datetime] = None

    job_status: str
    customer_name: Optional[str] = None
    customer_phone: Optional[str] = None

    class Config:
        from_attributes = True


class DriverHistoryResponse(BaseModel):
    tanker_id: int
    total: int
    items: list[DriverHistoryItem]

File: frontend/src/lib/history.ts
const API_BASE_URL =
    import.meta.env.VITE_API_BASE_URL?.replace(/\/+$/, "") || "http://127.0.0.1:8000";

export interface ClientHistoryItem {
    request_id: number;
    delivery_type: "batch" | "priority";
    request_status: string;
    volume_liters: number;
    created_at: string | null;
    completed_at: string | null;

    batch_id: number | null;
    member_id: number | null;
    batch_status: string | null;
    member_status: string | null;
    payment_status: string | null;
    refund_status: string | null;
    amount_paid: number | null;

    tanker_id: number | null;
    driver_name: string | null;

    delivery_id: number | null;
    delivery_status: string | null;
    planned_liters: number | null;
    actual_liters_delivered: number | null;
    otp_verified: boolean | null;
    delivered_at: string | null;
}

export interface ClientHistoryResponse {
    user_id: number;
    total: number;
    items: ClientHistoryItem[];
}

export interface DriverHistoryItem {
    job_type: "batch" | "priority";
    job_id: number;
    tanker_id: number;
    tanker_status: string | null;
    total_stops: number;
    delivered_stops: number;
    failed_stops: number;
    skipped_stops: number;
    total_planned_liters: number;
    total_actual_liters_delivered: number;
    started_at: string | null;
    completed_at: string | null;
    last_updated_at: string | null;
    job_status: string;
    customer_name: string | null;
    customer_phone: string | null;
}

export interface DriverHistoryResponse {
    tanker_id: number;
    total: number;
    items: DriverHistoryItem[];
}

async function apiGet<T>(path: string): Promise<T> {
    const response = await fetch(`${API_BASE_URL}${path}`);
    const data = await response.json();

    if (!response.ok) {
        throw new Error(data?.detail || "Request failed");
    }

    return data as T;
}

export function fetchClientHistory(userId: number) {
    return apiGet<ClientHistoryResponse>(`/history/users/${userId}`);
}

export function fetchDriverHistory(tankerId: number) {
    return apiGet<DriverHistoryResponse>(`/history/tankers/${tankerId}`);
}